// @solid/community-server/src/init/RuntimeConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuntimeConfig = void 0;
const Util_1 = require("../util/Util");
/**
 * This class holds all configuration options that can be defined by the user via the command line.
 *
 * Concretely, this contains data that is only relevant *after* dependency injection.
 */
class RuntimeConfig {
    constructor(data = {}) {
        this.reset(data);
    }
    reset(data) {
        var _a, _b, _c;
        this.pport = (_a = data.port) !== null && _a !== void 0 ? _a : 3000;
        this.pbase = Util_1.ensureTrailingSlash((_b = data.base) !== null && _b !== void 0 ? _b : `http://localhost:${this.port}/`);
        this.prootFilepath = (_c = data.rootFilepath) !== null && _c !== void 0 ? _c : process.cwd();
    }
    /**
     * The returned URL is ensured to have a trailing slash.
     */
    get base() {
        return this.pbase;
    }
    get port() {
        return this.pport;
    }
    get rootFilepath() {
        return this.prootFilepath;
    }
}
exports.RuntimeConfig = RuntimeConfig;
