// @solid/community-server/src/ldp/http/BasicTargetExtractor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicTargetExtractor = void 0;
const url_1 = require("url");
const TargetExtractor_1 = require("./TargetExtractor");
/**
 * Extracts an identifier from an incoming {@link HttpRequest}.
 * Uses URL library for basic parsing.
 * TODO: input requires more extensive cleaning/parsing based on headers (see #22).
 */
class BasicTargetExtractor extends TargetExtractor_1.TargetExtractor {
    async canHandle(input) {
        if (!input.url) {
            throw new Error('Missing URL.');
        }
        if (!input.headers.host) {
            throw new Error('Missing host.');
        }
    }
    async handle(input) {
        const isHttps = input.connection && input.connection.encrypted;
        const url = url_1.format({
            protocol: `http${isHttps ? 's' : ''}`,
            host: input.headers.host,
            pathname: input.url,
        });
        return { path: url };
    }
}
exports.BasicTargetExtractor = BasicTargetExtractor;
