// @solid/community-server/src/storage/conversion/QuadToTurtleConverter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuadToTurtleConverter = void 0;
const n3_1 = require("n3");
const ContentTypes_1 = require("../../util/ContentTypes");
const ConversionUtil_1 = require("./ConversionUtil");
const RepresentationConverter_1 = require("./RepresentationConverter");
/**
 * Converts `internal/quads` to `text/turtle`.
 */
class QuadToTurtleConverter extends RepresentationConverter_1.RepresentationConverter {
    async canHandle(input) {
        ConversionUtil_1.checkRequest(input, [ContentTypes_1.INTERNAL_QUADS], [ContentTypes_1.TEXT_TURTLE]);
    }
    async handle(input) {
        return this.quadsToTurtle(input.representation);
    }
    quadsToTurtle(quads) {
        const metadata = Object.assign(Object.assign({}, quads.metadata), { contentType: ContentTypes_1.TEXT_TURTLE });
        return {
            binary: true,
            data: quads.data.pipe(new n3_1.StreamWriter({ format: ContentTypes_1.TEXT_TURTLE })),
            metadata,
        };
    }
}
exports.QuadToTurtleConverter = QuadToTurtleConverter;
